### PR TITLE
NPE in Parquet Writer Metrics when data value max bound will overflow

### PR DESCRIPTION
**Pull request by szehon-ho**
A certain string in the input data (with a prefix of over 16 unparseable chars like high/low surrogates) triggered a NullPointerException in Parquet writer close:

```
java.lang.NullPointerException
		at org.apache.iceberg.parquet.ParquetUtil.toBufferMap(ParquetUtil.java:267)
		at org.apache.iceberg.parquet.ParquetUtil.footerMetrics(ParquetUtil.java:152)
		at org.apache.iceberg.parquet.ParquetUtil.footerMetrics(ParquetUtil.java:85)
		at org.apache.iceberg.parquet.ParquetWriter.metrics(ParquetWriter.java:151)
		at org.apache.iceberg.io.DataWriter.close(DataWriter.java:78)
		at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.closeCurrent(BaseTaskWriter.java:282)
		at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.close(BaseTaskWriter.java:298)
		at org.apache.iceberg.io.PartitionedWriter.close(PartitionedWriter.java:82)
		at org.apache.iceberg.io.BaseTaskWriter.abort(BaseTaskWriter.java:72)
		at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$6(WriteToDataSourceV2Exec.scala:447)
		at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1484)
```

The problem is that UnicodeUtil and BinaryUtil return null if fail to calculate a truncated upper bound for the string/binary (see UnicodeUtil.truncateStringMax() and BinaryUtil.truncateBinaryMax()).  A null value in the upperBound map then triggers a NPE in the ParquetUtil.toBufferMap class as it tries to call .getValue() on it.

```
  private static Map<Integer, ByteBuffer> toBufferMap(Schema schema, Map<Integer, Literal<?>> map) {
    Map<Integer, ByteBuffer> bufferMap = Maps.newHashMap();
    for (Map.Entry<Integer, Literal<?>> entry : map.entrySet()) {
      bufferMap.put(entry.getKey(),
          Conversions.toByteBuffer(schema.findType(entry.getKey()), entry.getValue().value()));
    }
    return bufferMap;
  }
```

### CODE REVIEW

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java [hunk 1]
```diff
@@ -112,5 +159,10 @@ public void testDataWriter() throws IOException {
     }
 
     Assert.assertEquals("Written records should match", records, writtenRecords);
+
+    Assert.assertTrue("Should have a valid lower bound", dataFile.lowerBounds().containsKey(1));
+    Assert.assertTrue("Should have a valid upper bound", dataFile.upperBounds().containsKey(1));
+    Assert.assertTrue("Should have a valid lower bound", dataFile.lowerBounds().containsKey(2));
```
Comment (szehon-ho) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java:165]: It seems UnicodeUtil and BinaryUtil truncateMin functions do not fail (they just truncate to the length and return).  

I hope it is ok to have a lowerBound but no upperBound for this column?

**Comment by szehon-ho** [comment]
@rdblue @aokolnychyi @RussellSpitzer  fyi for comments on whether this is the right fix for the issue, thanks!

**Comment by szehon-ho** [comment]
@aokolnychyi @RussellSpitzer could you guys take a look when you have some time?  Looks like the data is bad (can update the test case to make it more generic), but they are possible Parquet strings and Iceberg could be much more more graceful when getting such data

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java (hunk 1 above)
Comment (RussellSpitzer) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java:166]: I think we always do a null check before comparing, but lets check the metrics evaluators

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 2]
```diff
@@ -287,11 +287,19 @@ private static void increment(Map<Integer, Long> columns, int fieldId, long amou
...
+            Literal<CharSequence> truncatedMaxString = UnicodeUtil.truncateStringMax((Literal<CharSequence>) max,
+                truncateLength);
+            if (truncatedMaxString != null) {
+              upperBounds.put(id, truncatedMaxString);
+            }
             break;
           case FIXED:
           case BINARY:
-            upperBounds.put(id, BinaryUtil.truncateBinaryMax((Literal<ByteBuffer>) max, truncateLength));
+            Literal<ByteBuffer> truncatedMaxBinary = BinaryUtil.truncateBinaryMax((Literal<ByteBuffer>) max,
```
Comment (RussellSpitzer) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:298]: Do we need a test for this? Looks like we only got the unicode test?

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java (hunk 1 above)
Comment (RussellSpitzer) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java:166]: https://github.com/apache/iceberg/blob/bf9a22738b1160749b113f9b5f090a5e8d4c3827/api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java#L231-L237

https://github.com/apache/iceberg/blob/bf9a22738b1160749b113f9b5f090a5e8d4c3827/api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java#L256-L263 <- Yep should be safe

Comment (RussellSpitzer) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestParquetDataWriter.java:166]: Basically this means if the upper bound is missing (due to incalculable truncate) we just always fall back to "RowsMayMatch"

**Review by RussellSpitzer — COMMENTED** [review]
Ok so after a quick discussion I think we listed a few things
1. Test for binary max value, truncate a binary value which can't be incremented (1111)
2. Tests should check that upperBounds are set to null
3. Tests can focus just on triggering the truncate, remove the sort order stuff to make it clear what we are testing

After that I think we are good to go!

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java (hunk 2 above)
Comment (szehon-ho) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:298]: yep, will give a try

**Comment by szehon-ho** [comment]
@RussellSpitzer  thanks a lot for review, done.  The upper bound check in the test was already there (i clarified a bit the assert message to try to make it more clear).

**Comment by szehon-ho** [comment]
Renamed issue to reflect other use cases that will hit this, following deeper look during our conversation

**Review by RussellSpitzer — APPROVED** [review]
Ok this all looks good to me, let's merge it in

**Comment by RussellSpitzer** [comment]
Thanks @szehon-ho for the pr!

**Comment by szehon-ho** [comment]
Thanks for the review!

**Pull request merged**